quickwit-storage/src/bundle_storage.rs

```diff
@@ -0,0 +1,282 @@
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use crate::Storage;
+use crate::{StorageError, StorageResult};
+use async_trait::async_trait;
+use bytes::Bytes;
+use quickwit_common::HOTCACHE_FILENAME;
+use serde::Deserialize;
+use serde::Serialize;
+use std::collections::HashMap;
+use std::fmt;
+use std::fmt::Debug;
+use std::fs::{File, OpenOptions};
+use std::io::{self, ErrorKind, Write};
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use tantivy::common::BinarySerializable;
+use tantivy::HasLen;
+use tracing::error;
+
+/// Filename used for the bundle.
+pub const BUNDLE_FILENAME: &str = "bundle";
+
+/// BundleStorage bundles together multiple files into a single file.
+/// with some metadata
+pub struct BundleStorage {
+    storage: Arc<dyn Storage>,
+    bundle_file_name: PathBuf,
+    metadata: BundleStorageMetadata,
+}
+
+impl BundleStorage {
+    /// Creates a new BundleStorage.
+    ///
+    /// The provided data must include the footer_bytes at the end of the slice, but it can have more.
+    pub fn new(
+        storage: Arc<dyn Storage>,
+        bundle_file_name: PathBuf,
+        meta_data: &[u8],
+    ) -> io::Result<Self> {
+        let metadata = BundleStorageMetadata::open(meta_data)?;
+        Ok(BundleStorage {
+            storage,
+            bundle_file_name,
+            metadata,
+        })
+    }
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+struct BundleStorageMetadata {
+    pub(crate) files: HashMap<PathBuf, FileOffsets>,
+}
+
+impl BundleStorageMetadata {
+    fn open(data: &[u8]) -> io::Result<Self> {
+        let footer_size: u64 = BinarySerializable::deserialize(&mut &data[data.len() - 8..])?;
+
+        let footer_slice = &data[data.len() - 8 - footer_size as usize..data.len() - 8];
+        serde_json::from_slice(footer_slice)
+            .map_err(|err| io::Error::new(ErrorKind::InvalidData, err))
+    }
+    fn get(&self, path: &Path) -> StorageResult<FileOffsets> {
+        self.files
+            .get(path)
+            .cloned()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "not found in metadata").into())
+    }
+    fn exists(&self, path: &Path) -> bool {
+        self.files.contains_key(path)
+    }
+}
+
+#[async_trait]
+impl Storage for BundleStorage {
+    async fn put(&self, path: &Path, _payload: crate::PutPayload) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn copy_to_file(&self, path: &Path, _output_path: &Path) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn get_slice(&self, path: &Path, range: Range<usize>) -> crate::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        let new_range =
+            range.start + file_offsets.start as usize..range.end + file_offsets.start as usize;
```

Comment:
super nitpick... I'd commute the addition there.
Offset usually comes first.

No need to change if you disagree.

```
file_offsets.start as usize + range.start..file_offsets.start as usize +  range.end
```